pkg/kubelet/network/dns/dns.go

```diff
@@ -98,17 +104,35 @@ func (c *Configurer) getClusterDNS(pod *v1.Pod) ([]net.IP, error) {
 		networkName = arktosv1.NetworkDefault
 	}
 
-	if network, err := c.arktosV1Client.NetworksWithMultiTenancy(pod.Tenant).Get(networkName, metav1.GetOptions{}); err == nil {
-		if network.Status.Phase != arktosv1.NetworkReady {
-			return nil, fmt.Errorf("network %s/%s is in %q phase, not ready", network.Tenant, network.Name, network.Status.Phase)
+	if c.serviceLister == nil {
+		return nil, fmt.Errorf("failed to get serviceLister when getting service IP for network")
+	}
+	services, err := c.serviceLister.List(labels.Everything())
+	if err != nil {
+		return nil, fmt.Errorf("failed to list services when getting service IP for network")
+	}
+
+	const kubeDNSPrefix = "kube-dns-"
+	var kubeDNSServiceName = kubeDNSPrefix + networkName
+
+	// Get all services and find out the cluster IP of service kube-dns-{network}
+	for i := range services {
+		service := services[i]
+
+		// ignore services where ClusterIP is "None" or empty
+		if !v1helper.IsServiceIPSet(service) {
+			continue
 		}
 
-		ip := net.ParseIP(network.Status.DNSServiceIP)
-		if ip != nil {
-			return []net.IP{ip}, nil
+		// Get the cluster IP of service kube-dns-{network} as name server
+		// inside the file /etc/resolv.conf of pod
+		if service.Tenant == pod.Tenant && service.Name == kubeDNSServiceName {
+			ip := net.ParseIP(service.Spec.ClusterIP)
+			if ip != nil {
+				klog.V(4).Infof("Pod NAME: %q | ClusterDNS IP : %q | networkName : %q | Tenant : %q ", pod.Name, ip,  networkName, pod.Tenant)
+				return []net.IP{ip}, nil
```

Comment:
it seems that on success, it always returns one result. If that is the case, is there any other reason for the interface method to return collection instead of singular one?

Reply:
The community code at https://github.com/CentaurusInfra/arktos/blob/8e5c308a2a1da21e18a8949db7224d16fba01f08/pkg/kubelet/network/dns/dns.go#L386-L387 call the method 'getClusterDNS' and return collection of IP other than singular one. Just follow-up community method and do not change too many codes.

		var clusterDNS []net.IP
		if clusterDNS, err = c.getClusterDNS(pod); err != nil {